### PR TITLE
Improve docs for BinaryData.ToString()

### PR DESCRIPTION
**Pull request by JoshLove-msft**

### CODE REVIEW

**Comment by GrabYourPitchforks** [comment]
In general I don't have a concern with clarifying docs, but it does strike me as a bit weird to call out JSON string behavior specifically. Why are JSON strings promoted to a first-class citizen within the remarks section, but JSON objects/arrays aren't mentioned, and weirdness resulting from `<arbitary binary data>.ToString()` not called out? Is this doc change motivated by seeing how people are using / misusing this API in practice? And if that's the case, is a code analyzer a better alternative to doc updates?

**Comment by JoshLove-msft** [comment]
> In general I don't have a concern with clarifying docs, but it does strike me as a bit weird to call out JSON string behavior specifically. Why are JSON strings promoted to a first-class citizen within the remarks section, but JSON objects/arrays aren't mentioned, and weirdness resulting from `<arbitary binary data>.ToString()` not called out? Is this doc change motivated by seeing how people are using / misusing this API in practice? And if that's the case, is a code analyzer a better alternative to doc updates?

Yes, this is motivated by users not understanding how to get a JSON string and being confused by the output from ToString. I'm not sure how an analyzer would work in general here as we would not be able to tell whether it is a valid ToString usage or not. The main goal is to just introduce the idea of ToStringFromJson<string>, as it isn't obvious to users that they would need to do this.

**Comment by GrabYourPitchforks** [comment]
> we would not be able to tell whether it is a valid ToString usage or not.

@JoshLove-msft What would a "valid" `ToString` usage look like for this type? From my understanding of the type, this type is a literally a container for arbitrary binary data. That would imply that the only "correct" way to convert this to a human-readable representation is for the caller to signal what the data represents and thus what transform is valid: JSON unwrapping, UTF-8 conversion, hex conversion (for binary data), etc. I'm almost tempted to suggest marking the `ToString` method as `[Obsolete]` in addition to putting in the doc comment.

@bartonjs Do you have thoughts on this? In recent reviews you've mentioned concerns about users calling `ToString` on various exchange types while harboring incorrect assumptions about just what the `ToString` method does.

**Review by GrabYourPitchforks — APPROVED** [review]
Approved to unblock this since it's an improvement over the current situation. But per feedback in the issue conversation, I think relying solely on docs is the wrong long-term solution, and we need to think about more in-your-face mechanisms to steer users toward correct behaviors.

**Comment by JoshLove-msft** [comment]
> > we would not be able to tell whether it is a valid ToString usage or not.
> 
> @JoshLove-msft What would a "valid" `ToString` usage look like for this type? From my understanding of the type, this type is a literally a container for arbitrary binary data. That would imply that the only "correct" way to convert this to a human-readable representation is for the caller to signal what the data represents and thus what transform is valid: JSON unwrapping, UTF-8 conversion, hex conversion (for binary data), etc. I'm almost tempted to suggest marking the `ToString` method as `[Obsolete]` in addition to putting in the doc comment.
> 

ToString does UTF-8 conversion, so any UTF-8 encoded bytes would be valid. It isn't necessarily invalid to call ToString on JSON, it is just probably not want the user wants in most cases, since there are special JSON deserialization methods on BinaryData. A case where it makes sense to use ToString would be when you are using some textual represenation, e.g. plaintext, or XML. In those cases, ToString makes sense. I agree that it is somewhat ambiguous, and that this is something of a usability bandaid.

File: src/libraries/System.Memory.Data/src/System/BinaryData.cs [hunk 1]
```diff
@@ -192,6 +192,11 @@ public static BinaryData FromObjectAsJson<T>(T jsonSerializable, JsonSerializerO
         /// <summary>
         /// Converts the value of this instance to a string using UTF-8.
         /// </summary>
+        /// <remarks>
+        /// If the underlying object is a JSON string, calling <see cref="ToString"/> will retain the wrapping double quotes in the
+        /// resulting string. If you want to deserialize the JSON string to a string, thereby removing the wrapping double quotes,
+        /// call <see cref="ToObjectFromJson{String}"/> instead.
```
Comment (bartonjs) [inline src/libraries/System.Memory.Data/src/System/BinaryData.cs:198]: Sharing Levi's "this feels too special-cased", I think something more like

```suggestion
        /// No special treatment is given to the contents of the data, it is merely decoded as a UTF-8 string.
        /// For a JPEG or other binary file format the string will largely be nonsense with many embedded NUL characters,
        /// and UTF-8 JSON values will look like their file/network representation,
        /// including starting and stopping quotes on a string.
```

And then add a `<seealso cref="ToObjectFromJson{String}" />` after the `</returns>`

**Review by adamsitnik — APPROVED** [review]
I've applied all the suggestions and pushed them directly to @JoshLove-msft branch. LGTM

**Comment by adamsitnik** [comment]
Failures are unrelated (#56894), merging

**Pull request merged**